Gloo-only CPU-based monitored barrier

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54742 test_c10d: use with_nccl_blocking_wait decorator
* #54741 test_c10d: Run tests with nccl_async_error_handling
* #54740 Provide a decorator to set/unset nccl blocking wait for tests
* #54558 [NCCL][Blocking Wait] Log set exceptions when checking for exceptions in
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**

Closes https://github.com/pytorch/pytorch/issues/52876

Implements a barrier by doing send/recv to rank 0, and rank 0 waits for these requests and on timeout, throws an exception indicating which rank did not join in the given timeout.

This barrier is only intended for CPU use cases and built into process group gloo, and will be used for debugging synchronization/hang issues.

Differential Revision: [D26921357](https://our.internmc.facebook.com/intern/diff/D26921357/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26921357/)!

rohan-varma committed:
Gloo-only CPU-based monitored barrier

Closes https://github.com/pytorch/pytorch/issues/52876

Implements a barrier by doing send/recv to rank 0, and rank 0 waits for these requests and on timeout, throws an exception indicating which rank did not join in the given timeout.

This barrier is only intended for CPU use cases and built into process group gloo, and will be used for debugging synchronization/hang issues.

Differential Revision: [D26921357](https://our.internmc.facebook.com/intern/diff/D26921357/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26921357/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Gloo-only CPU-based monitored barrier"


Closes https://github.com/pytorch/pytorch/issues/52876

Implements a barrier by doing send/recv to rank 0, and rank 0 waits for these requests and on timeout, throws an exception indicating which rank did not join in the given timeout.

This barrier is only intended for CPU use cases and built into process group gloo, and will be used for debugging synchronization/hang issues.

Differential Revision: [D26921357](https://our.internmc.facebook.com/intern/diff/D26921357/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26921357/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Gloo-only CPU-based monitored barrier"


Closes https://github.com/pytorch/pytorch/issues/52876

Implements a barrier by doing send/recv to rank 0, and rank 0 waits for these requests and on timeout, throws an exception indicating which rank did not join in the given timeout.

This barrier is only intended for CPU use cases and built into process group gloo, and will be used for debugging synchronization/hang issues.

Differential Revision: [D26921357](https://our.internmc.facebook.com/intern/diff/D26921357/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26921357/)!

[ghstack-poisoned]

zhaojuanmao commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5119,3 +5119,30 @@ def test_output_unused_in_loss(self):
+                with self.assertRaisesRegex(RuntimeError, f"Rank {failed_rank}"):
+                    process_group.monitored_barrier(timeout)
+            else:
+                # Other ranks will report standard gloo error, only rank 0 knows
```
what if all reduce has error in rank 0?

rohan-varma replied:
The main use case here is to report hangs/issues for non-zero ranks. Most users read master (rank 0) logs first, so issues on rank 0 should be pretty apparent, the challenge is usually hangs in nonzero ranks. 

I think one small enhancement here could be to report a better error when nonzeroro ranks timeout on send/recv (currently we just let it crash with gloo error), to indicate something may have been crashed in rank 0.

zhaojuanmao reviewed (commented):
great work! left some comments and questions mostly

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1075,7 +1075,17 @@ that adds a prefix to each key inserted to the store.
+              py::call_guard<py::gil_scoped_release>())
+          .def(
+              "monitored_barrier",
+              []( const c10::intrusive_ptr<::c10d::ProcessGroup>& self, double timeoutSeconds) {
```
Maybe we just use `timedelta` in the API so it automatically gets converted to std::chrono::... via pybind?

rohan-varma replied:
For RPC framework, I think we ended up going with float for the timeout over timedelta, should we keep it consistent here?

rohan-varma replied:
Looks like float is also needed to make this work with TorchScript. There is a separate TODO to make TorchBind work with std::chrono (https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/init.cpp#L1599), so we can probably scope this PR to only add python support and add torchscript support as a follow up.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -106,6 +109,19 @@ namespace c10d {
+  auto elapsedTime = std::chrono::steady_clock::now() - startTime;
+  auto elapsedMillis = timeout -
+      std::chrono::duration_cast<std::chrono::milliseconds>(elapsedTime);
+  auto remaining = std::max(elapsedMillis.count(), static_cast<long>(1));
```
Not sure why we return 1 here if there is no remaining time? Shouldn't it be 0?

rohan-varma replied:
Wanted to provide a small buffer of 1 ms so that we call into the recv/send collective, and then it will likely timeout and report the appropriate error. If that seems hacky, we can change this to 0.

pritamdamania87 replied:
I don't think we should rely on having a buffer of 1ms to ensure something is called, in different environments the 1ms buffer may or may not work.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -2661,4 +2677,73 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupGloo::barrier(
+      try {
+        recvWork->wait(remainingTime);
+      } catch (const std::exception& e) {
+        LOG(ERROR) << "Rank " << rank << " failed to pass monitoredBarrier in "
```
nit: Can create a message string and reuse in LOG and exception.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -2661,4 +2677,73 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupGloo::barrier(
+    // Rank 0 recv
+    if (rank != 0) {
+      auto recvWork = recv(t, rank, t1);
+      auto remainingTime = monitoredBarrierTimeout == kUnsetTimeout
```
Why do we use `getRemainingtime` here? We can just pass `monitoredBarrierTimeout` to `wait`?

rohan-varma replied:
Do you mean we can omit the call to `getRemainingTime` for the very first call to the collective, or in general why do we use `getRemainingTime` at all? 

For the latter, I implemented it this way so that if timeout = 100s for example, then we wait a total of 100s and not 100s for each call to a collective work's wait(). For the former point, I guess we can omit the call for the very first collective since all the time is remaining.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -2661,4 +2677,73 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupGloo::barrier(
+  }
+  auto startTime = std::chrono::steady_clock::now();
+  auto worldSize = this->getSize();
+  for (int rank = 0; rank < worldSize; ++rank) {
```
Might be better to have one for loop, where you do send and recv for all ranks and then a second for loop that waits for all the work. That way you don't have this pattern of send-wait for each individual rank.

pritamdamania87 commented on torch/lib/c10d/test/ProcessGroupGlooTest.cpp:
```diff
@@ -426,6 +426,43 @@ void testBarrier(const std::string& path) {
           std::move(event_lists), GLOO_STR, size, allShapes, /* verify_shapes */ false);
 }
 
+void testMonitoredBarrier(const std::string& path) {
```
Why do we have a cpp test instead of just a python test?

rohan-varma replied:
We could test it all through python, but I figured it's better to have a cpp test for better coverage and since we are planning to call this API from C++ directly once it is part of processgroup wrapper.

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5119,3 +5119,30 @@ def test_output_unused_in_loss(self):
+            else:
+                # Other ranks will report standard gloo error, only rank 0 knows
+                # ranks that failed to respond to barrier.
+                try:
+                    process_group.monitored_barrier(timeout)
+                except RuntimeError as e:
+                    pass
```
nit: Use self.assertRaises(RuntimeError)?

rohan-varma commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1075,7 +1075,17 @@ that adds a prefix to each key inserted to the store.
+              py::call_guard<py::gil_scoped_release>())
+          .def(
+              "monitored_barrier",
+              []( const c10::intrusive_ptr<::c10d::ProcessGroup>& self, double timeoutSeconds) {
```
We use float in RPC for timeout, I think we explicitly decided to go with float over datetime.timedelta when designing timeout for rpc framework, so thought it would be good to be consistent here. Will change it to float.

rohan-varma replied:
Actually looks like we can't change it to float type since torchscript type system only supports double, and c10d APIs are now registered through torchbind:

```
caffe2/aten/src/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h:196:5: error: static_assert failed due to requirement 'guts::false_t<float>::value' "You tried to register a kernel with an unsupported input type: float. Please use double instead."
```

rohan-varma commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -2661,4 +2677,73 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupGloo::barrier(
+    // Rank 0 recv
+    if (rank != 0) {
+      auto recvWork = recv(t, rank, t1);
+      auto remainingTime = monitoredBarrierTimeout == kUnsetTimeout
```
If user does not pass in timeout, then timeout is `kUnsetTimeout`. If we pass in `kUnsetTimeout` to collective comm. APIs, the default timeout specified in init_process_group ends up being used.

rohan-varma commented on torch/lib/c10d/test/ProcessGroupGlooTest.cpp:
```diff
@@ -426,6 +426,43 @@ void testBarrier(const std::string& path) {
+  for (auto & t : threads) {
+    t.join();
+  }
+  // Failure case: Only rank 0 calls into monitored barrier, should result in error
```
Will add this test, it should pretty much behave the same if the rank hanging is not 0.

rohan-varma replied:
Added in python test

rohan-varma committed:
Update on "Gloo-only CPU-based monitored barrier"


Closes https://github.com/pytorch/pytorch/issues/52876

Implements a barrier by doing send/recv to rank 0, and rank 0 waits for these requests and on timeout, throws an exception indicating which rank did not join in the given timeout.

This barrier is only intended for CPU use cases and built into process group gloo, and will be used for debugging synchronization/hang issues.

Differential Revision: [D26921357](https://our.internmc.facebook.com/intern/diff/D26921357/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26921357/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Gloo-only CPU-based monitored barrier"


Closes https://github.com/pytorch/pytorch/issues/52876

Implements a barrier by doing send/recv to rank 0, and rank 0 waits for these requests and on timeout, throws an exception indicating which rank did not join in the given timeout.

This barrier is only intended for CPU use cases and built into process group gloo, and will be used for debugging synchronization/hang issues.

Differential Revision: [D26921357](https://our.internmc.facebook.com/intern/diff/D26921357/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26921357/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Gloo-only CPU-based monitored barrier"


Closes https://github.com/pytorch/pytorch/issues/52876

Implements a barrier by doing send/recv to rank 0, and rank 0 waits for these requests and on timeout, throws an exception indicating which rank did not join in the given timeout.

This barrier is only intended for CPU use cases and built into process group gloo, and will be used for debugging synchronization/hang issues.

Differential Revision: [D26921357](https://our.internmc.facebook.com/intern/diff/D26921357/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26921357/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Gloo-only CPU-based monitored barrier"


Closes https://github.com/pytorch/pytorch/issues/52876

Implements a barrier by doing send/recv to rank 0, and rank 0 waits for these requests and on timeout, throws an exception indicating which rank did not join in the given timeout.

This barrier is only intended for CPU use cases and built into process group gloo, and will be used for debugging synchronization/hang issues.

Differential Revision: [D26921357](https://our.internmc.facebook.com/intern/diff/D26921357/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26921357/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Gloo-only CPU-based monitored barrier"


Closes https://github.com/pytorch/pytorch/issues/52876

Implements a barrier by doing send/recv to rank 0, and rank 0 waits for these requests and on timeout, throws an exception indicating which rank did not join in the given timeout.

This barrier is only intended for CPU use cases and built into process group gloo, and will be used for debugging synchronization/hang issues.

Differential Revision: [D26921357](https://our.internmc.facebook.com/intern/diff/D26921357/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26921357/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Gloo-only CPU-based monitored barrier"


Closes https://github.com/pytorch/pytorch/issues/52876

Implements a barrier by doing send/recv to rank 0, and rank 0 waits for these requests and on timeout, throws an exception indicating which rank did not join in the given timeout.

This barrier is only intended for CPU use cases and built into process group gloo, and will be used for debugging synchronization/hang issues.

Differential Revision: [D26921357](https://our.internmc.facebook.com/intern/diff/D26921357/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26921357/)!

[ghstack-poisoned]

rohan-varma commented:
@pritamdamania87 Actually, I think we need to keep the pattern of rank 0 doing:

for each rank:
recv() ; wait()

for each rank:
send() ; wait()

Otherwise, if we do the current version of the PR:
for each rank:
    recv() ; send()

for each rank:
    waitrecv()
    waitsend()

then there is a risk that some ranks will be unblocked and make it out of the barrier, while ranks stuck somewhere else would not. It would be good to maintain the invariant that all ranks make it out or not (imo), so I'll revert this to the older version

rohan-varma committed:
Update on "Gloo-only CPU-based monitored barrier"

blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**

Closes https://github.com/pytorch/pytorch/issues/52876

Implements a barrier by doing send/recv to rank 0, and rank 0 waits for these requests and on timeout, throws an exception indicating which rank did not join in the given timeout.

This barrier is only intended for CPU use cases and built into process group gloo, and will be used for debugging synchronization/hang issues.

Differential Revision: [D26921357](https://our.internmc.facebook.com/intern/diff/D26921357/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26921357/)!

[ghstack-poisoned]

zhaojuanmao commented:
The codes look good to me. But I'm a little confused right now about the purpose of monitored_barrier() after looking at the tests. 

I think what we want for these cases are follows, and I feel monitored_barrier() did not give such ability based on my current understanding. Please let me know if I misunderstood. 

**Case1, every rank call all_reduce, but ranks hang and timedout because there are mismatched tensors among ranks.** 
for this case, monitored_barrier() only gives info that other ranks timed out, did not provide info about why it timed out and thus help debugging. 

**Case2,  rank1 did not enter all_reduce, all other ranks enter all_reduce. All ranks will hang and timed out**
How could monitored_barrier() call help for this case. all ranks have already timed out before entering monitored_barrier(). 

even for your test case, rank0 did not enter all reduce and caused other ranks timed out. monitored_timeout() only gives info that other ranks timed out, did not provide info about other ranks timed out is due to rank0 did not enter the all_reduce, and thus can not provide info to help debugging.  ==> Am I understanding correctly?

pritamdamania87 commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -2662,4 +2682,65 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupGloo::barrier(
+  // out first, bringing down the job without reporting which rank timed out.
+  if (rank != 0) {
+    auto sendWork = send(commTensor, 0, t1);
+    sendWork->wait();
```
nit: you can wait below after kicking off `recvWork`?

pritamdamania87 commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -106,6 +110,22 @@ namespace c10d {
+  auto elapsedTime = std::chrono::steady_clock::now() - startTime;
+  auto remainingMillis = timeout -
+      std::chrono::duration_cast<std::chrono::milliseconds>(elapsedTime);
+  const int kMinTimeThresh = 1;
```
Oh are we returning 1 as the minimum value here since `wait()` assumes a value of 0 as no timeout? If so, I think we could just rename this function to `checkAndGetRemainingTime` and just throw an exception from here that the barrier timed out.

rohan-varma replied:
Sounds good, I think it might be cleaner to return something like `-1` in this case and let monitoredBarrier() impl take care of raising the error?

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5119,3 +5119,72 @@ def test_output_unused_in_loss(self):
+            for _ in range(10):
+                process_group.allreduce(tensors).wait()
+            # Run monitored barrier
+            timeout = timedelta(seconds=2)
```
nit: maybe we can use `timeout=0.1` everywhere to ensure the test runs faster?

rohan-varma replied:
Sounds good, though if we use too small of a timeout, monitored_barrier() in cases where it's expected to succeed may fail, since there is some latency in the interprocess comm.

pritamdamania87 commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5119,3 +5119,72 @@ def test_output_unused_in_loss(self):
+            os.environ["NCCL_BLOCKING_WAIT"] = "1"
+            nccl_pg = dist.new_group(
+                ranks=list(i for i in range(int(self.world_size))),
+                timeout=timedelta(seconds=10),
```
Can we make this timeout lower? I'm guessing the allreduce below will block for 10s as a result of this?

rohan-varma replied:
Sure, might be okay to do a 1-2s timeout as I want to ensure the below allreduce that's intended to be successful does not timeout, and we can pass in a short timeout to wait() to ensure a quick test time.

rohan-varma commented:
> **Case1, every rank call all_reduce, but ranks hang and timedout because there are mismatched tensors among ranks.**
> for this case, monitored_barrier() only gives info that other ranks timed out, did not provide info about why it timed out and thus help debugging.

To clarify, if all ranks are stuck in allreduce, then monitored_barrier isn't really intended to help too much here since no rank will ever call it, and we'd just have stuckness. In the case that rank 0 calls monitored_barrier, and some other ranks are stuck in allreduce, it will report a stuck rank. In the case that nonzero rank calls into monitored_barrier, it will eventually timeout, maybe we can add logging/notification here that rank 0 failed to respond to the monitored barrier as a follow up diff.


> 
> **Case2, rank1 did not enter all_reduce, all other ranks enter all_reduce. All ranks will hang and timed out**
> How could monitored_barrier() call help for this case. all ranks have already timed out before entering monitored_barrier().
> 

In this case, rank 0 should report rank 1 (or some other rank that didn't enter barrier) as an error when crashing in monitored_barrier. monitored_barrier() here will be helpful since user will be able to track down a stuck rank from rank 0, and go to those rank's logs to help debug further.


> even for your test case, rank0 did not enter all reduce and caused other ranks timed out. monitored_timeout() only gives info that other ranks timed out, did not provide info about other ranks timed out is due to rank0 did not enter the all_reduce, and thus can not provide info to help debugging. ==> Am I understanding correctly?

Yes, this is correct, monitored_barrier is not intended to give information such as other ranks timed out in allreduce. The main goal here is to report a rank that failed to enter the barrier, and this gives the user additional debug information (a faulty rank). The goal is to avoid the user from having to search through the logs of hundreds of ranks, and only have to look at rank 0 to get clues about faulty ranks.

rohan-varma committed:
Update on "Gloo-only CPU-based monitored barrier"

blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**

Closes https://github.com/pytorch/pytorch/issues/52876

Implements a barrier by doing send/recv to rank 0, and rank 0 waits for these requests and on timeout, throws an exception indicating which rank did not join in the given timeout.

This barrier is only intended for CPU use cases and built into process group gloo, and will be used for debugging synchronization/hang issues.

Differential Revision: [D26921357](https://our.internmc.facebook.com/intern/diff/D26921357/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26921357/)!

[ghstack-poisoned]

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5119,3 +5119,88 @@ def test_output_unused_in_loss(self):
+            if world_size == 2:
+                err_regex = "Rank 1"
+            else:
+                err_regex = f"Rank [1-{world_size - 1}]"
```
This regex is because for world_size > 2, it is possible for any nonzero rank to timeout out. We hold a unordered_map of the work objects, so the implementation doesn't enforce a specific order of checking the ranks. Will add a follow up change to provide an option to wait for all failed ranks, though.

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5119,3 +5119,88 @@ def test_output_unused_in_loss(self):
+            os.environ["NCCL_BLOCKING_WAIT"] = "1"
+            nccl_pg = dist.new_group(
+                ranks=list(i for i in range(int(self.world_size))),
+                timeout=timedelta(seconds=2),
```
We use a smaller timeout in the allreduce that we expect to fail, however this one should be generous enough for the allreduce before that which we want to succeed.

rohan-varma committed:
Update on "Gloo-only CPU-based monitored barrier"

blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**

Closes https://github.com/pytorch/pytorch/issues/52876

Implements a barrier by doing send/recv to rank 0, and rank 0 waits for these requests and on timeout, throws an exception indicating which rank did not join in the given timeout.

This barrier is only intended for CPU use cases and built into process group gloo, and will be used for debugging synchronization/hang issues.

Differential Revision: [D26921357](https://our.internmc.facebook.com/intern/diff/D26921357/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26921357/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Gloo-only CPU-based monitored barrier"

blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**

Closes https://github.com/pytorch/pytorch/issues/52876

Implements a barrier by doing send/recv to rank 0, and rank 0 waits for these requests and on timeout, throws an exception indicating which rank did not join in the given timeout.

This barrier is only intended for CPU use cases and built into process group gloo, and will be used for debugging synchronization/hang issues.

Differential Revision: [D26921357](https://our.internmc.facebook.com/intern/diff/D26921357/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26921357/)!

[ghstack-poisoned]

zhaojuanmao commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -2662,4 +2692,98 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupGloo::barrier(
+  std::vector<int> processedRanks;
+
+  // Wait for send/recv from all ranks
+  for (auto& work : works) {
```
This is to identify first rank that did not pass monitored barrier, right?  as you mentioned, the follow up will be to find out all the ranks that did not pass monitored barriers, as we do not know who is the real first rank that did not pass monitored barrier, right?

rohan-varma replied:
This is to identify a general rank that did not pass the barrier, not necessarily the first. We don't enforce an order on checking for the ranks, should we do this?

zhaojuanmao replied:
right, I think my comment is not clear. I meant it will identify the first stuck rank in the order how works are ordered in the map or in the order when works are waited.  Did you check whether this order is always deterministic or random?

rohan-varma replied:
I guess it will be random because unordered_map doesn't give any guarantee on element order. I think we should just convert it to use std::map here and then have a proper order (least to greatest rank), no. of ranks is not that big that there will be a slowdown.

zhaojuanmao commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -2662,4 +2692,98 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupGloo::barrier(
+      auto remainingTime = getRemainingTime(startTime, monitoredBarrierTimeout);
+      if (remainingTime.count() < 0) {
+        std::string rankInfo;
+        if (processedRanks.size() > 0) {
+          rankInfo = c10::str(
+              "Successfully processed ranks: ",
+              c10::Join(", ", processedRanks));
```
should not it also include this rank info and saying current rank=work.first did not pass the monitored barrier?

rohan-varma replied:
The rank logging this information is always rank 0, and I'm not sure if we can say current rank = work.first did not pass the monitored barrier.

In this codepath, we basically ran out of time during processing and did not call into `work.wait()`. The rank may indeed have been healthy but we ran out of time to process it, so we cannot surely say either way and we need to fail here due to the timeout.

In the general case, we fail in `work.wait()` where our timeout is exhausted and that is handled below, where we do log the failed rank.

zhaojuanmao replied:
I see, that makes sense!

zhaojuanmao reviewed (approved):
looks good. please feel free to add follow up change to provide an option to wait for all failed ranks.

rohan-varma committed:
Update on "Gloo-only CPU-based monitored barrier"

blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**

Closes https://github.com/pytorch/pytorch/issues/52876

Implements a barrier by doing send/recv to rank 0, and rank 0 waits for these requests and on timeout, throws an exception indicating which rank did not join in the given timeout.

This barrier is only intended for CPU use cases and built into process group gloo, and will be used for debugging synchronization/hang issues.

Differential Revision: [D26921357](https://our.internmc.facebook.com/intern/diff/D26921357/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26921357/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Gloo-only CPU-based monitored barrier"

blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**

Closes https://github.com/pytorch/pytorch/issues/52876

Implements a barrier by doing send/recv to rank 0, and rank 0 waits for these requests and on timeout, throws an exception indicating which rank did not join in the given timeout.

This barrier is only intended for CPU use cases and built into process group gloo, and will be used for debugging synchronization/hang issues.

Differential Revision: [D26921357](https://our.internmc.facebook.com/intern/diff/D26921357/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26921357/)!

[ghstack-poisoned]

rohan-varma commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -2662,4 +2692,96 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupGloo::barrier(
+        auto error = c10::str(kNoRemainingTimeError, "\n", rankInfo);
+        logAndThrow(error, error);
+      }
+      try {
+        c10::intrusive_ptr<ProcessGroup::Work> commWork;
+        // C++ does not support std::get<i> for variable i.
+        if (i == 0) {
+          commWork = std::get<0>(work.second);
+        } else {
+          commWork = std::get<1>(work.second);
+        }
+        commWork->wait(remainingTime);
```
Sure, we also need to check and see if we need to throw error between successive calls to getRemainingTime and wait, but we should be able to do this without too much code duplication.

rohan-varma committed:
Update on "Gloo-only CPU-based monitored barrier"

blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**

Closes https://github.com/pytorch/pytorch/issues/52876

Implements a barrier by doing send/recv to rank 0, and rank 0 waits for these requests and on timeout, throws an exception indicating which rank did not join in the given timeout.

This barrier is only intended for CPU use cases and built into process group gloo, and will be used for debugging synchronization/hang issues.

Differential Revision: [D26921357](https://our.internmc.facebook.com/intern/diff/D26921357/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26921357/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Gloo-only CPU-based monitored barrier"

blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**
blocking wait
* #54557 [NCCL] Enhance watchdog to log exceptions
* #54220 [DDP] Use monitored barrier in DDP
* #54219 [DDP] Remove redundant pass statement
* #53787 Expose dist.monitored_barrier() API
* **#53773 Gloo-only CPU-based monitored barrier**

Closes https://github.com/pytorch/pytorch/issues/52876

Implements a barrier by doing send/recv to rank 0, and rank 0 waits for these requests and on timeout, throws an exception indicating which rank did not join in the given timeout.

This barrier is only intended for CPU use cases and built into process group gloo, and will be used for debugging synchronization/hang issues.

Differential Revision: [D26921357](https://our.internmc.facebook.com/intern/diff/D26921357/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D26921357/)!

[ghstack-poisoned]
